### PR TITLE
pkg/proc: support swiss table map implementation

### PR DESCRIPTION
**Pull request by aarzilli**
Adds support for the swiss table implementation for Go's maps.

### CODE REVIEW

**Comment by aarzilli** [comment]
The swissmap backend has been enabled by default on 1.24 so I think we can review this now.

File: pkg/proc/mapiter.go [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+			if it.tab == nil {
+				return false
+			}
+			if it.tab.index != it.dirIdx {
+				it.nextTable()
+				continue
+			}
+		}
+
+		for ; it.groupIdx <= uint64(it.tab.groups.Len); it.nextGroup() {
```
Comment (prattmic) [inline pkg/proc/mapiter.go:469]: it.tab.groups.Len is the array length, right? Thus shouldn't this be < not <=?

**Review by prattmic — COMMENTED** [review]
I took a look at the swiss implementation (not the classic one). LGTM, the only thing I saw that looks actually incorrect is the comment about <= on it.tab.groups.Len.

File: pkg/proc/variables.go [hunk 2]
```diff
@@ -1992,12 +1991,7 @@ func (v *Variable) loadMap(recurseLevel int, cfg LoadConfig) {
 	errcount := 0
 	for it.next() {
 		key := it.key()
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/proc/variables.go:1993]: We didn't handle either. I missed it when I first implemented it years ago and nobody ever reported it as a bug.
I've added support to both with this PR, for classic map is the code in `(*mapIteratorClassic).key()` that does the `k.Kind == reflect.Ptr && !it.keyTypeIsPtr`, for swiss maps it's in `next`, the if for `it.curKey.Kind == reflect.Ptr && !it.keyTypeIsPtr` (and their equivalents for value).

File: pkg/proc/mapiter.go [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+
+				var err1, err2 error
+				it.curKey, err1 = cur.toField(it.slotFieldKey)
+				it.curValue, err2 = cur.toField(it.slotFieldElem)
+				if err1 != nil || err2 != nil {
+					it.v.Unreadable = fmt.Errorf("error accessing swiss map slot: %v %v", err1, err2)
+					return false
+				}
+
+				if it.curKey.Kind == reflect.Ptr && !it.keyTypeIsPtr {
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/proc/mapiter.go:499]: The map type is no longer stored in the map object, if I'm not mistaken to get to the map type I would have to use the DW_AT_go_runtime_type. It's very annoying.

File: pkg/proc/mapiter.go (hunk 1 above)
Comment (aarzilli) [reply pkg/proc/mapiter.go:469]: You're right.

File: pkg/proc/mapiter.go [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+					it.curKey = it.curKey.maybeDereference()
+				}
+				if it.curValue.Kind == reflect.Ptr && !it.elemTypeIsPtr {
+					it.curValue = it.curValue.maybeDereference()
+				}
+
+				it.slotIdx++
+				if it.slotIdx >= uint32(it.group.slots.Len) {
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/proc/mapiter.go:507]: Done.

File: pkg/proc/mapiter.go [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+		return
+	}
+
+	if it.dirLen <= 0 {
+		// small maps, convert it.dirPtr to be of type *group, then dereference it
+		it.dirPtr.DwarfType = pointerTo(fakeArrayType(1, it.groupType), it.v.bi.Arch)
+		it.dirPtr.RealType = it.dirPtr.DwarfType
+		it.dirPtr = it.dirPtr.maybeDereference()
+		it.dirLen = 1
+		it.tab = &swissTable{groups: it.dirPtr} // so that we don't try to load this later on
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/proc/mapiter.go:441]: Cross-process memory reads are slow anyway.

**Review by derekparker — APPROVED** [review]
LGTM

**Pull request merged**